kotlin/reflect/src/main/java/com/squareup/moshi/kotlin/reflect/KotlinJsonAdapter.kt

```diff
@@ -248,9 +254,7 @@ class KotlinJsonAdapterFactory : JsonAdapter.Factory {
 
     for (parameter in constructor.parameters) {
       val binding = bindingsByName.remove(parameter.name)
-      if (binding == null && !parameter.isOptional) {
-        throw IllegalArgumentException("No property for required constructor ${parameter}")
-      }
+      require(!(binding == null && !parameter.isOptional)) { "No property for required constructor $parameter" }
```

Comment:
```suggestion
      require(binding != null || parameter.isOptional) { "No property for required constructor $parameter" }
```

Reply:
Will fix in a follow-up, didn't know you were doing a pass before merging 😬